### PR TITLE
tests: Add further testing of xdsl_opt_main

### PR DESCRIPTION
**Pull request by webmiche**
This PR adds several tests for edge cases in xdsl_opt_main.

### CODE REVIEW

File: tests/xdsl_opt/test_xdsl_opt.py [hunk 1]
```diff
@@ -1,15 +1,100 @@
 from xdsl.xdsl_opt_main import xDSLOptMain
 from contextlib import redirect_stdout
 from io import StringIO
+import pytest
```
Comment (georgebisbas) [inline tests/xdsl_opt/test_xdsl_opt.py:4]: Better move them to the top of the file

File: tests/xdsl_opt/test_xdsl_opt.py [hunk 2]
```diff
@@ -1,15 +1,100 @@
...
 def test_opt():
-    filename = 'tests/xdsl_opt/empty_program.xdsl'
-    opt = xDSLOptMain(args=[filename])
+    opt = xDSLOptMain(args=[])
     assert list(opt.available_frontends.keys()) == ['xdsl', 'mlir']
     assert list(opt.available_targets.keys()) == ['xdsl', 'irdl', 'mlir']
     assert list(opt.available_passes.keys()) == []
 
+
+def test_empty_program():
```
Comment (georgebisbas) [inline tests/xdsl_opt/test_xdsl_opt.py:15]: Maybe the following three tests can be parameterised

**Review by georgebisbas — COMMENTED** [review]
Do we need these empty files?

File: tests/xdsl_opt/test_xdsl_opt.py [hunk 3]
```diff
@@ -1,15 +1,100 @@
...
+    inp = open(filename_in, 'r').read()
+    opt.run()
+    expected = open(filename_out, 'r').read()
+
+    assert inp.strip() == expected.strip()
+
+    open(filename_out, 'w').write("")
+
+
+def test_wrong_target():
```
Comment (math-fehr) [inline tests/xdsl_opt/test_xdsl_opt.py:63]: We should change the `xdsl_opt_main` to change target depending on the output file name.

File: tests/xdsl_opt/test_xdsl_opt.py [hunk 4]
```diff
@@ -1,15 +1,100 @@
...
+
+    assert e.value.args[0] == "Unrecognized pass: wrong"
+
+
+def test_wrong_file_extension():
+    filename = 'tests/xdsl_opt/empty_program.wrong'
+    opt = xDSLOptMain(args=[filename])
+
+    f = StringIO("")
+    with pytest.raises(Exception) as e:
```
Comment (math-fehr) [inline tests/xdsl_opt/test_xdsl_opt.py:87]: Maybe this should fail at creation of `xDSLOptMain`, not sure.
We can do this in another PR anyway

File: tests/xdsl_opt/test_xdsl_opt.py [hunk 5]
```diff
@@ -1,15 +1,100 @@
...
+        opt.run()
+
+    assert e.value.args[0] == "Unrecognized file extension 'wrong'"
+
+
+def test_coverage():
+    filename = 'tests/xdsl_opt/empty_program.xdsl'
+    opt = xDSLOptMain(
+        args=[filename, '--generate-coverage', '--exec-root=' + os.getcwd()])
```
Comment (math-fehr) [inline tests/xdsl_opt/test_xdsl_opt.py:96]: Can we check that the coverage was created?

**Review by math-fehr — CHANGES_REQUESTED** [review]
Nice! I added some comments.

File: tests/xdsl_opt/test_xdsl_opt.py (hunk 2 above)
Comment (webmiche) [reply tests/xdsl_opt/test_xdsl_opt.py:15]: I am not quite sure what you mean. You mean using a fancy `@pytest.mark.parametrize`? I am unfortunately not very knowledge about the possibilities this provides. Could you educate me on how it could help in this case?

Comment (webmiche) [reply tests/xdsl_opt/test_xdsl_opt.py:15]: I think I understood now. let me try!

Comment (webmiche) [reply tests/xdsl_opt/test_xdsl_opt.py:15]: I implemented something. Would be great to get more feedback!

**Review by georgebisbas — APPROVED** [review]

**Review by math-fehr — APPROVED** [review]
Looks good!
